### DISCUSSION TITLE
Opinions collected! cyfs-gateway can become an independent product！

### DISCUSSION THREAD

**Discussion by waterflier** [buckyos/buckyos#56] [Ideas]
Everyone: 

We have accumulated extensive experience from the CYFS project, including the design of the next-generation HTTP protocol, the comprehensive design and implementation of a P2P protocol, and the separation of Named Data transmission and logical data transmission based on NDN network theory. However, one of our most significant (and painful) lessons is the importance of releasing products that are accessible to the general public. Like many other network protocol projects, CYFS faces a cold start problem: to fully utilize CYFS, one must leave the current browser/server ecosystem and go through extensive basic environment configuration, leading to CYFS environments often only running in laboratories. Additionally, the vast self-developed infrastructure poses challenges to system stability.

Therefore, I hope that on the path to achieving our goals with BuckyOS, we can release products that "target users can genuinely use." On the one hand, we should prioritize using mature existing technologies (e.g., using GlusterFS for DFS initially) over performance and elegance to achieve key goals. On the other hand, we should aim for "components that can be independently productized" as much as possible, allowing these products to have their own goals and independent release capabilities, thereby reducing the overall integration complexity of the system.

Based on this premise, I believe that at least two standalone products can be extracted from the current BuckyOS:

CYFS Protocol Stack as an Extensible Proxy Server Framework: From a certain perspective, the CYFS protocol stack is an extensible proxy server framework. By leveraging mature proxy protocols, it allows client/server compatibility with new network protocols. Considering the concept of an 'extensible proxy server,' we can identify many widespread needs: transparent proxy, privacy access for specific sites based on rule engines, accelerated access, parental control over children's internet behavior, home VPN (connecting back to the home LAN while traveling), etc. We can release a superior home SDN software based on the CYFS protocol stack (we could start discussions with cyfs-gateway, and perhaps retain the product name Swallow?).

BuckyOS Backup Suite: Given our strong focus on data reliability, we can initially separate the scalable backup system included in BuckyOS into a standalone software (BuckyOS Backup Suite?), enabling a third-party Backup Server ecosystem that complies with the BuckyOS Backup Target protocol to be independent of the BuckyOS release.

Among these two standalone products, the demand for the backup suite is relatively straightforward, and I have already organized it. The requirements for the cyfs-gateway, however, are much more diverse. What kind of super proxy software do people hope to use? (This software may be installed on smart routers.) I would like to hear everyone's opinions.

**Comment by wugren** [comment]
Although the gateway needs to be productized independently, the functionalities of the product should all be usable by buckyos.

According to the network design of buckyos, core nodes within the same zone need to be able to connect to each other, such as etcd and glusterfs nodes. To be compatible with these third-party software, implementing a VPN is a good solution.

For cross-zone access, an SDN should be implemented. Nodes within a zone should have the option to join the SDN, and only nodes that have joined the SDN can allow cross-zone access.

**Comment by waterflier** [comment]
To implement VLANs, it seems necessary for the cyfs-gateway to function as an "IP-Gateway." This requires the cyfs-gateway to operate in a special environment. However, if the application that needs to be managed is running in Docker, there might be a simpler way to achieve the following scenario:

- After the cyfs-gateway starts, it creates a virtual VLAN.
- A process A is started using Docker, and through configuration, process A is effectively added to the VLAN.
- On Windows, by using common VPN software and entering the configuration information generated by the cyfs-gateway, the Windows machine also joins the VLAN.

### Detailed Steps:

1. **Creating a Virtual VLAN with cyfs-gateway**:
    - The cyfs-gateway will initialize and set up a virtual VLAN.
    - Ensure the VLAN configuration is properly defined, including VLAN ID and network settings.

2. **Starting a Docker Process and Joining the VLAN**:
    - Launch process A in Docker with network settings configured to join the virtual VLAN.
    - This involves using Docker's network capabilities to connect containers to the VLAN created by cyfs-gateway.
    - For example, we can create a Docker network that matches the VLAN settings and use it to start process A:
      ```bash
      docker network create --driver bridge --subnet 192.168.100.0/24 vlan_network
      docker run --network vlan_network --name process_a my_app_image
      ```

3. **Connecting Windows to the VLAN via VPN**:
    - Use a common VPN software on Windows.
    - Enter the configuration information generated by the cyfs-gateway to connect the Windows machine to the VLAN.
    - This configuration information should include the necessary VPN settings (e.g., server address, authentication details, and network routes).

### Example Configuration for Docker:

- **Create a Docker Network**:
  ```bash
  docker network create --driver bridge --subnet 192.168.100.0/24 vlan_network
  ```

- **Start a Docker Container**:
  ```bash
  docker run --network vlan_network --name process_a my_app_image
  ```

### Example VPN Configuration for Windows:

1. **Generate VPN Configuration on cyfs-gateway**:
   - Configure the cyfs-gateway to provide VPN settings.
   - Export the VPN configuration file (e.g., `.ovpn` file for OpenVPN).

2. **Configure VPN on Windows**:
   - Install a VPN client (e.g., OpenVPN).
   - Import the configuration file generated by the cyfs-gateway.
   - Connect to the VPN using the imported settings.

By following these steps, we can effectively create a VLAN environment with the cyfs-gateway acting as the IP-Gateway, enabling Docker processes and Windows machines to join the VLAN seamlessly.

**Comment by waterflier** [comment]
I have compiled some of the futures of cyfs-gateway that I thought of, and we can add to them.

## SOCKS5 Proxy with Rule-Based Tunnel Connections

1. **IP Privacy Protection**:
   - Hide the client's real IP address from the target host by routing traffic through a SOCKS5 proxy.

2. **Connection Acceleration**:
   - Improve connection speed by choosing an appropriate intermediate tunnel when the client cannot connect directly to the target host or experiences slow connections.

3. **Rule Engine**:
   - Implement a rule engine that allows automatic selection of the appropriate tunnel based on conditions, typically different domain names.
   - Design the rule engine to be user-friendly, similar to iptables but easier for ordinary users to understand. These rules can be easily generated via ChatGPT.

4. **Extensible Tunnel System**:
   - Use a plugin system to support various tunnels, allowing easy expansion and customization.

## cyfs-gateway is DNS server

1. **DNS Query via Tunnel/HTTPS**:
   - Enforce rules that require certain domain names to use tunnels or HTTPS for DNS queries to prevent domain pollution.

2. **Host Configuration Support**:
   - Enable host configuration, facilitating the testing of systems on devices that cannot configure hosts, such as iOS devices.

3. **Embedded Scripting for Custom Logic**:
   - Use embedded scripting (e.g., Lua) to extend the DNS resolution logic with simple customizations.

4. **Plugin System for Different Query Backends**:
   - Support various query backends through a plugin system, including native support for decentralized domain name systems like ENS.

## P2P(BDT) Reverse Proxy for NAT Traversal

1. **Dynamic DNS (DDNS) Needs**:
   - Handle basic DDNS requirements by dynamically refreshing an IP address associated with a domain name when port mapping has exposed internal network services.

2. **Reverse Proxy Logic**:
   - Allow users to configure their VPS with cyfs-gateway, mapping specific ports to internal network devices' device IDs and ports (e.g., 80/443). This configuration will enable external access to internal HTTP servers via a public domain name.

3. **Port Mapping with P2P**:
   - Set up local port forwarding on a Windows machine running cyfs-gateway, such as mapping local port 13389 to a device ID:3389 within the internal network. This configuration allows remote desktop clients to access the internal PC by connecting to 127.0.0.1:13389.

**Comment by wugren** [comment]
Socks5 Proxy and reverse Proxy are not good implementation methods, as demonstrated in the demo. 

The experience of using Socks5 Proxy with third-party software is very poor. It only supports client connections and requires client support to function. If the client does not support it, it cannot be used, making it essentially impossible to implement for systems like etcd or glusterfs. 

Additionally, the reverse proxy implementation method also cannot support issues related to dynamic ports.

**Comment by waterflier** [comment]
Our experience in DEMO is that these requirements can only be solved by SDN/VPN technology, and the related requirements are new to cyfs-gateway. In this issue, we need to expand the specific requirements in DEMO into general requirements and determine the boundaries of functions.

I listed the requirements of traditional "smart gateway" as I understand them, not just these. We definitely plan to integrate SDN/VPN related requirements

Thank you for your comments~
